src/core/attributes.rs

```diff
@@ -130,8 +131,10 @@ where
     pub fn on_flush(self) {
         let gauge = self.metric;
         let op = self.operation;
-        write_lock!(self.target.get_attributes().flush_listeners)
-            .push(Arc::new(move |now| gauge.write(op(now), Labels::default())));
+        let mut listeners = write_lock!(self.target.get_attributes().flush_listeners);
+        if !listeners.contains_key(gauge.metric_id()) {
+            listeners.insert(gauge.metric_id().clone(), Arc::new(move |now| gauge.write(op(now), Labels::default())));
+        }
```

Comment:
This behavior is the exact opposite of the one we need. It's "preserve existing" but we need "replace existing by this new one". What about to make it configurable with an enum with `panic`, `preserve`, `replace`?

Our use case is monitoring of a data source that is reloaded each few minutes. The life cycle of the wrapper with monitoring is the same as of the inner object.

```rust
/// Wrapper for a blacklist that adds monitoring capabilities.
pub(crate) struct Monitored<B> {
    inner: B,

    matches: Counter,
    cleans: Counter,
    reloads: Counter,
    records: Gauge,
    broken_records: Gauge,
    memory: Gauge,
    // TODO: Optionally monitor also
    // - reloads that failed
    // - last modified time
}

impl<B: Blacklist> Monitored<B> {
    pub(crate) fn new(monitor: &Monitor, inner: B, stats: LoadingStats, memory: usize) -> Self {
        let me = Monitored {
            inner,
            matches: monitor.counter("matches").init(),
            cleans: monitor.counter("cleans").init(),
            reloads: monitor.counter("reloads").init(),
            broken_records: monitor.gauge("broken-records").init(),
            records: monitor.gauge("records").init(),
            memory: monitor.gauge("memory").init(),
        };

        me.reloads.count(1);
        me.records.value(stats.records);
        me.broken_records.value(stats.broken_records);
        me.memory.value(memory);

        me
    }
}
```

Reply:
And just to be sure, the second gauge is returned from a "cache" so g1 and g2 point to the same instance, right?

```rust
let g1 = metrics.gauge("uptime")
let g2 = metrics.gauge("uptime")
```

Reply:
Yup, cache returns same instance. Actually, flush listener's behavior to preserve oldest instance comes from the cache. We could also replace the existing entry, but the result to the client should be the same as both instances should behave the same way and be the metric for all intents and purposes.

Reply:
I may be wrong but I don't think it will behave the same at all, the results may be different.

```rust
    pub fn on_flush(self) {
        let gauge = self.metric;
        let op = self.operation;
        let mut listeners = write_lock!(self.target.get_attributes().flush_listeners);
        if !listeners.contains_key(gauge.metric_id()) {
            listeners.insert(gauge.metric_id().clone(), Arc::new(move |now| gauge.write(op(now), Labels::default())));
        }
    }
```

The key point is that lambdas capture their environment. If the `listeners` already contains that key, `self.operation` will be lost. If the following code is called for for the same metric e.g. after few minutes, it will still report the age based on the "oldest now" and not based on the "latest now".

```rust
let now = Instant::now();
bucket.observe(bucket.gauge("age"), move |_| now.elapsed().as_micros() as MetricValue).on_flush();
```rust

Reply:
Oh I see! This is pretty derpy of me. I'll fix it.